### PR TITLE
core/state: trie prefetcher change: calling trie() doesn't stop the associated subfetcher

### PR DESCRIPTION
**Pull request by jwasinger**
This pulls trie prefetcher changes from https://github.com/ethereum/go-ethereum/pull/29029 into a separate PR.

It changes the trie prefetcher s.t. a call to `trie` returns a trie (after any pending tasks have been resolved in the subfetcher).  After a call to `trie`, the subfetcher is now available for more work (instead of terminating).

### CODE REVIEW

File: core/state/trie_prefetcher.go
Comment (rjl493456442) [inline core/state/trie_prefetcher.go:170]: looks like this field is not really used.

File: core/state/trie_prefetcher.go
Comment (rjl493456442) [inline core/state/trie_prefetcher.go:259]: It might be possible that the prefetching loop is already terminated, but a new task is scheduled.

It doesn't satisfy our requirement, isn't it? We should somehow return a flag saying: the given task won't be executed as the prefetcher is already terminated?

File: core/state/trie_prefetcher.go
Comment (rjl493456442) [inline core/state/trie_prefetcher.go:124]: Once the trie is retrieved, the corresponding subPrefetcher is terminated.

It's problematic for the blocks before ByzantiumFork as the trie is retrieved after each transaction, instead of at the block boundary.

Comment (rjl493456442) [reply core/state/trie_prefetcher.go:124]: I feel like we should make it as a block function. If there are some ongoing prefetching tasks, the function will be blocked until these tasks are fully executed.

File: core/state/trie_prefetcher.go
Comment (jwasinger) [reply core/state/trie_prefetcher.go:259]: Afaict, we will only ever call `triePrefetcher.prefetch` before `triePrefetcher.close` has been called.  and they are not invoked concurrently.  so this case you bring up shouldn't happen (?).

File: core/state/trie_prefetcher.go
Comment (jwasinger) [reply core/state/trie_prefetcher.go:126]: Done in the latest commit.

File: core/state/trie_prefetcher.go
Comment (jwasinger) [inline core/state/trie_prefetcher.go:233]: One assumption that I have with this PR is that opening the trie after starting a subfetcher should not fail.  In any case where we start a subfetcher, we have already ensured that it has an existing trie in the db:
* calls to prefetch storage slots can only happen if the account trie root is non-empty
* the root trie will never be empty (it's a pathological case that can be ignored?).

So if my assumptions here are correct, then we should panic when this fails.  Otherwise, if I am wrong and it is a valid case, a subfetcher will exit `loop` and any subsequent calls to `schedule`, `close`, `trie` will hang indefinitely.

**Comment by jwasinger** [comment]
Failed tests... the last commit borks something.

File: core/state/trie_prefetcher.go
Comment (rjl493456442) [reply core/state/trie_prefetcher.go:259]: right

File: core/state/trie_prefetcher.go
Comment (rjl493456442) [inline core/state/trie_prefetcher.go:114]: I think we should mention that: once the `triePrefetcher` is closed, all the following tasks scheduled will not be executed.

File: core/state/trie_prefetcher.go
Comment (rjl493456442) [inline core/state/trie_prefetcher.go:129]: it might be possible that the subFetcher is present while the subFetcher.Trie is nil(e.g. failed to open it). `Nil` should be returned in this case.

e.g.
 
```go
if fetcher == nil || fetcher.trie == nil {
    return nil
}
```

File: core/state/trie_prefetcher.go
Comment (rjl493456442) [reply core/state/trie_prefetcher.go:233]: It's possible that trie opening is failing because of state missing. 

For an instance, the node is started and required to rollback to a historic block, therefore all the available states are discarded. However, the prefetcher of pending block is already created *before the rollback operation*, all the trie opening will be failed.

It's pretty easy to repro it by running `debug.setHead(oldBlock)` in the path mode.

**Review by rjl493456442 — APPROVED** [review]

File: core/state/trie_prefetcher.go
Comment (holiman) [inline core/state/trie_prefetcher.go:117]: Is there ever a multi-thread accessing prefetch / close ? If so, might be better to make `closed` into an `atomic.Bool`? But if we don't access it that way, it's fine.

**Review by holiman — COMMENTED** [review]
Could you please elaborate a bit more about what this PR tries to achieve, and why that is better. 
As in, why it's needed for your upcoming work, but also what the effects on the _current_ code is. Does it improve anything? Simplify something? Degrade something?

**Comment by rjl493456442** [comment]
The biggest difference this pull request introduces is:

Originally, when a trie is requested, the prefetching tasks are suspended and resumed once the live trie is copied and returned. This approach could cause problems for witness collection. For instance:

If slot A is accessed but not modified, the relevant trie nodes should be wrapped as the witness. Subsequently, slot A will be scheduled for prefetching and queued in the task list.

However, if the corresponding storage trie is retrieved without slot A being preloaded at that time, then the relevant trie nodes won't be included in the witness.

In this pull request, the prefetching tasks scheduled before retrieving the trie will always be executed. Thus the issue is resolved.

File: core/state/trie_prefetcher.go
Comment (rjl493456442) [reply core/state/trie_prefetcher.go:117]: Prefetcher is manipulated in the statedb, which is not regarded as thread-safe. I believe we never do multi-thread accessing prefetch.

**Review by holiman — COMMENTED** [review]
Looking good, but we need to run it on the benchmarking infra for a bit

**Comment by jwasinger** [comment]
Unfortunately, the added overhead to block processing and increased memory usage means that this PR is a non-starter.

**Comment by holiman** [comment]
I think it's maybe a bit premature to close this. Please add the relevant graphs/charts from the benchmark-run, so we can discuss and reason about it

**Comment by holiman** [comment]
```
Mar 14 04:09:55 bench05.ethdevops.io geth INFO [03-14|03:09:55.583] Starting peer-to-peer node instance=Geth/v1.14.0-unstable-3c26ffeb/linux-amd64/go1.21.8
Mar 14 04:10:12 bench06.ethdevops.io geth INFO [03-14|03:10:11.942] Starting peer-to-peer node instance=Geth/v1.14.0-unstable-287cead1/linux-amd64/go1.21.8
```
So apparently `bench06` is running this PR.  

Screenshot 2024-03-15 at 10-27-28 Dual Geth - Grafana

And `bench06` is markedly slower, mainly in the bucket called `execution`. I guess that's where the time spent waiting for the subfetchers to finish is counted. 

The prefetchers fetch roughly the same amount of data (bench06 a bit more slots): 
Screenshot 2024-03-15 at 10-35-54 Dual Geth - Grafana

And the both have same "misses", that is, work performed that was never needed (e.g. fetching slots at the end of transaction, but the next transaction set the slots back to the previous value). 

Screenshot 2024-03-15 at 10-36-45 Dual Geth - Grafana

In the "optimistic" prefetcher, if we overload it with requests, it will just fetch what it can, and then we'll say "ok stop now". In the strict one, overloading it with requests means we'll have to wait it out, it will penalize use. So if we do this, maybe we have to be more clever about what to schedule. 

I'm not really 100% sure what scenarios causes the wasted work, or the delivery misses. Might be worth investigating a bit

File: core/state/trie_prefetcher.go
Comment (karalabe) [inline core/state/trie_prefetcher.go:197]: This change seems problematic. Whereas the original code allowed to schedule the request and then fly off continuing execution, the new code (along with L217 below) will block until a previous request is done (actually, block on the second request).

Any particular reason you made the async notifier synchronous?

**Comment by karalabe** [comment]
There's a change in this PR that converted the async wake notifier into a sync one, which can potentially have a really heavy hit on performance as it ill block execution during preloading. Perhaps that's the root of the slowdown.

File: core/state/trie_prefetcher.go
Comment (holiman) [reply core/state/trie_prefetcher.go:197]: Well spotted!

**Comment by karalabe** [comment]
Closing and reopening my own because OP didn't permit maintainer changes.

**Comment by jwasinger** [comment]
:'( I don't know why this is suddenly something I have to explicitly enable.  Until recently, PRs I've made in the past have been maintainer-modifiable by default....

**Pull request closed without merging**
